platform/platform-migratable/src/main/java/org/codice/ddf/platform/migratable/impl/PlatformMigratable.java

```diff
@@ -138,14 +161,81 @@ public void doImport(ImportMigrationContext context) {
         .stream()
         .map(context::getEntry)
         .forEach(me -> me.restore(false));
+    importKeystores(context);
+    importServiceWrapperFiles(context);
+  }
+
+  @Override
+  public void doVersionUpgradeImport(ImportMigrationContext context, String migratableVersion) {
+    if (Float.parseFloat(migratableVersion) > Float.parseFloat(getVersion())) {
+      context
+          .getReport()
+          .record(
+              new MigrationException(
+                  IMPORT_UNSUPPORTED_MIGRATABLE_VERSION_ERROR,
+                  migratableVersion,
+                  getId(),
+                  getVersion()));
+      return;
+    }
+
+    LOGGER.debug("Upgrading system properties...");
+    upgradeCustomSystemProperties(context);
+
+    LOGGER.debug("Importing remaining system files...");
+    PlatformMigratable.UPGRADEABLE_SYSTEM_PATHS
+        .stream()
+        .map(context::getEntry)
+        .forEach(me -> me.restore(false));
+    importKeystores(context);
+    importServiceWrapperFiles(context);
+  }
+
+  private void upgradeCustomSystemProperties(ImportMigrationContext context) {
+    Properties importedProps = new Properties();
+    Properties currentProps;
+
+    try {
+      File fileOnSystem =
+          Paths.get(System.getProperty("ddf.home"))
+              .toRealPath(LinkOption.NOFOLLOW_LINKS)
+              .resolve(Paths.get("etc", "custom.system.properties"))
+              .toFile();
+
+      importedProps.load(
+          context
+              .getEntry(PlatformMigratable.CUSTOM_SYSTEM_PROPERTIES_PATH)
+              .getInputStream()
+              .orElseThrow(IOException::new));
+      currentProps = new Properties(fileOnSystem);
+
+      currentProps
+          .entrySet()
+          .stream()
+          .filter(
+              e ->
+                  !importedProps.containsKey(e.getKey())
+                      || SYSTEM_PROPERTIES_TO_PRESERVE.contains(e.getKey()))
+          .forEach(e -> importedProps.put(e.getKey(), e.getValue()));
+
+      importedProps.save(fileOnSystem);
+    } catch (IOException e) {
+      LOGGER.warn("Failed to upgrade system properties", e);
```

Comment:
❓ Shouldn't we record something in `context.getReport()`?

Reply:
Yes, I think we should.